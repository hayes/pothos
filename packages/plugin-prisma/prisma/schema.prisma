// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema
datasource db {
  provider = "sqlite"
  url      = "file:./dev.db"
}

generator client {
  provider = "prisma-client-js"
  output   = "../tests/client"
}

generator pothos {
  provider          = "node -r @swc-node/register ./src/generator.ts"
  clientOutput      = "../tests/client"
  pluginPath        = "../src"
  output            = "../tests/generated.ts"
  generateDatamodel = true
  documentation     = true
}

/// Model for Posts!
model Post {
  id        Int         @id @default(autoincrement())
  bigIntId  BigInt      @unique
  createdAt DateTime    @unique @default(now())
  updatedAt DateTime    @updatedAt
  /// Title for the post
  title     String
  /// Content for the post
  content   String?
  /// Has this post been published?
  published Boolean     @default(false)
  author    User        @relation(fields: [authorId], references: [id])
  comments  Comment[]
  authorId  Int
  media     PostMedia[]
  views     Decimal     @default(0)
}

model Media {
  id           Int         @id @default(autoincrement())
  url          String
  posts        PostMedia[]
  uploadedBy   User?       @relation(fields: [uploadedById], references: [id])
  uploadedById Int?
}

model PostMedia {
  id      Int   @id @default(autoincrement())
  post    Post  @relation(fields: [postId], references: [id])
  media   Media @relation(fields: [mediaId], references: [id])
  postId  Int
  mediaId Int
  order   Int

  @@unique([postId, mediaId])
}

model Comment {
  id        Int      @id @default(autoincrement())
  createdAt DateTime @unique @default(now())
  content   String
  author    User     @relation(fields: [authorId], references: [id])
  post      Post     @relation(fields: [postId], references: [id])
  authorId  Int
  postId    Int
}

model Profile {
  id     Int     @id @default(autoincrement())
  bio    String?
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model User {
  id        Int       @id @default(autoincrement())
  email     String    @unique
  name      String?
  posts     Post[]
  comments  Comment[]
  profile   Profile?
  followers Follow[]  @relation("followers")
  following Follow[]  @relation("following")
  Media     Media[]
}

model Follow {
  fromId Int
  toId   Int
  from   User @relation("following", fields: [fromId], references: [id])
  to     User @relation("followers", fields: [toId], references: [id])

  @@id([fromId, toId], name: "compositeID")
}

model Unrelated {
  id   Int     @id @default(autoincrement())
  name String?
}

model WithID {
  id                  String                @id
  FindUniqueRelations FindUniqueRelations[]
}

model WithUnique {
  id                  String                @unique
  FindUniqueRelations FindUniqueRelations[]
}

model WithCompositeID {
  a String
  b String

  FindUniqueRelations FindUniqueRelations[]

  @@id([a, b])
}

model WithCompositeUnique {
  a String
  c String? @unique
  b String

  FindUniqueRelations FindUniqueRelations[]

  @@unique([a, c])
  @@unique([a, b])
}

model FindUniqueRelations {
  id String @id

  withID_id String
  withID    WithID @relation(fields: [withID_id], references: [id])

  withUnique_id String
  withUnique    WithUnique @relation(fields: [withUnique_id], references: [id])

  withCompositeID_a String
  withCompositeID_b String
  withCompositeID   WithCompositeID @relation(fields: [withCompositeID_a, withCompositeID_b], references: [a, b])

  withCompositeUnique_a String
  withCompositeUnique_b String
  withCompositeUnique   WithCompositeUnique @relation(fields: [withCompositeUnique_a, withCompositeUnique_b], references: [a, b])
}
