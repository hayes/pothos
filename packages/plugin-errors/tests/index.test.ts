import SchemaBuilder from '@pothos/core';
import { execute, printSchema, subscribe } from 'graphql';
import { gql } from 'graphql-tag';
import ErrorPlugin from '../src';
import { builder, builderWithCustomErrorTypeNames } from './example/builder';
import { createSchema } from './example/schema';

const schema = createSchema(builder);

describe('errors plugin', () => {
  it('generates expected schema', () => {
    expect(printSchema(schema)).toMatchSnapshot();

    expect(() => {
      builder.toSchema();
    }).not.toThrow();
  });

  it('query some stuff', async () => {
    const query = gql`
      query {
        simpleError {
          __typename
          ... on QuerySimpleErrorResult {
            data
          }
          ... on Error {
            message
          }
        }
        extendedError {
          __typename
          ... on QueryExtendedErrorResult {
            data
          }
          ... on Error {
            message
          }
        }

        simpleErrorError: simpleError(throw: true) {
          __typename
          ... on QuerySimpleErrorResult {
            data
          }
          ... on Error {
            message
          }
        }
        extendedErrorError: extendedError(throw: "error") {
          __typename
          ... on QueryExtendedErrorResult {
            data
          }
          ... on Error {
            message
          }
        }

        extendedErrorExtended: extendedError(throw: "extended") {
          __typename
          ... on QueryExtendedErrorResult {
            data
          }
          ... on Error {
            message
          }
        }

        extendedErrorExtended2: extendedError(throw: "extended2") {
          __typename
          ... on QueryExtendedErrorResult {
            data
          }
          ... on Error {
            message
          }
        }

        extendedErrorOther: extendedError(throw: "other") {
          __typename
          ... on QueryExtendedErrorResult {
            data
          }
          ... on Error {
            message
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: {},
    });

    expect(result).toMatchInlineSnapshot(`
      {
        "data": {
          "extendedError": {
            "__typename": "QueryExtendedErrorSuccess",
            "data": "ok",
          },
          "extendedErrorError": {
            "__typename": "BaseError",
            "message": "Error from extendedError",
          },
          "extendedErrorExtended": {
            "__typename": "ExtendedError",
            "message": "Error from extendedError",
          },
          "extendedErrorExtended2": {
            "__typename": "Extended2Error",
            "message": "Error from extendedError",
          },
          "extendedErrorOther": null,
          "simpleError": {
            "__typename": "QuerySimpleErrorSuccess",
            "data": "ok",
          },
          "simpleErrorError": {
            "__typename": "BaseError",
            "message": "Error from simpleError field",
          },
        },
        "errors": [
          [GraphQLError: Unexpected error value: { message: "Error from extendedError" }],
        ],
      }
    `);
  });

  it('query directResult', async () => {
    const query = gql`
      query {
        directResult {
          __typename
          ... on DirectResult {
            id
          }
        }
        withError: directResult(shouldThrow: true) {
          __typename
          ... on Error {
            message
          }
        }
      }
    `;

    const result = await execute({
      schema,
      document: query,
      contextValue: {},
    });

    expect(result).toMatchInlineSnapshot(`
      {
        "data": {
          "directResult": {
            "__typename": "DirectResult",
            "id": "123",
          },
          "withError": {
            "__typename": "BaseError",
            "message": "Boom",
          },
        },
      }
    `);
  });

  it('supports generating custom names', () => {
    const schemaWithCustomTypeNames = createSchema(builderWithCustomErrorTypeNames);
    expect(printSchema(schemaWithCustomTypeNames)).toMatchSnapshot();

    expect(() => {
      builderWithCustomErrorTypeNames.toSchema();
    }).not.toThrow();
  });

  it('subscription without errors', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "data": {
            "test": {
              "__typename": "SubscriptionTestSuccess",
              "data": 1,
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "SubscriptionTestSuccess",
              "data": 2,
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "SubscriptionTestSuccess",
              "data": 3,
            },
          },
        },
      ]
    `);
  });

  it('subscription null', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test(returnNull: true) {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    // biome-ignore lint/style/noUnusedTemplateLiteral: snapshot
    expect(results).toMatchInlineSnapshot(`[]`);
  });

  it('subscription error on subscribe', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test(errorOnSubscribe: true) {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on subscribe",
            },
          },
        },
      ]
    `);
  });

  it('subscription error in resolve', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test(errorOnResolve: true) {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on resolve",
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on resolve",
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on resolve",
            },
          },
        },
      ]
    `);
  });

  it('subscription error in iterable', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test(errorInIterable: true) {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "data": {
            "test": {
              "__typename": "SubscriptionTestSuccess",
              "data": 1,
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on subscribe",
            },
          },
        },
      ]
    `);
  });

  it('subscription error in iterable and resolve', async () => {
    const result = await subscribe({
      schema,
      document: gql`
        subscription {
          test(errorInIterable: true, errorOnResolve: true) {
            __typename
            ... on Error {
              message
            }
            ... on SubscriptionTestResult {
              data
            }
          }
        }
      `,
    });

    const results = [];
    for await (const value of result as AsyncIterable<{ data: { hello: number } }>) {
      results.push(value);
    }

    expect(results).toMatchInlineSnapshot(`
      [
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on resolve",
            },
          },
        },
        {
          "data": {
            "test": {
              "__typename": "BaseError",
              "message": "error on subscribe",
            },
          },
        },
      ]
    `);
  });
});

describe('onResolvedError callback', () => {
  it('calls onResolvedError when errors are handled', async () => {
    const resolvedErrors: Error[] = [];

    const testBuilder = new SchemaBuilder<{}>({
      plugins: [ErrorPlugin],
      errors: {
        defaultTypes: [],
        onResolvedError: (error) => {
          resolvedErrors.push(error);
        },
      },
    });

    testBuilder.objectType(Error, {
      name: 'Error',
      fields: (t) => ({
        message: t.exposeString('message'),
      }),
    });

    testBuilder.queryType({
      fields: (t) => ({
        testField: t.field({
          type: 'String',
          errors: {
            types: [Error],
          },
          resolve: () => {
            throw new Error('Test error message');
          },
        }),
      }),
    });

    const testSchema = testBuilder.toSchema();

    const result = await execute({
      schema: testSchema,
      document: gql`
        query {
          testField {
            __typename
            ... on Error {
              message
            }
          }
        }
      `,
      contextValue: {},
    });

    expect(resolvedErrors).toHaveLength(1);
    expect(resolvedErrors[0]).toBeInstanceOf(Error);
    expect(resolvedErrors[0].message).toBe('Test error message');

    expect(result).toMatchInlineSnapshot(`
      {
        "data": {
          "testField": {
            "__typename": "Error",
            "message": "Test error message",
          },
        },
      }
    `);
  });

  it('does not call onResolvedError for unhandled errors', async () => {
    const resolvedErrors: Error[] = [];

    class HandledError extends Error {
      constructor(message: string) {
        super(message);
        this.name = 'HandledError';
      }
    }

    class UnhandledError extends Error {
      constructor(message: string) {
        super(message);
        this.name = 'UnhandledError';
      }
    }

    const testBuilder = new SchemaBuilder<{}>({
      plugins: [ErrorPlugin],
      errors: {
        defaultTypes: [],
        onResolvedError: (error) => {
          resolvedErrors.push(error);
        },
      },
    });

    testBuilder.objectType(HandledError, {
      name: 'HandledError',
      fields: (t) => ({
        message: t.exposeString('message'),
      }),
    });

    testBuilder.queryType({
      fields: (t) => ({
        testField: t.field({
          type: 'String',
          errors: {
            types: [HandledError],
          },
          resolve: () => {
            throw new UnhandledError('This error is not handled by the plugin');
          },
        }),
      }),
    });

    const testSchema = testBuilder.toSchema();

    const result = await execute({
      schema: testSchema,
      document: gql`
        query {
          testField {
            __typename
          }
        }
      `,
      contextValue: {},
    });

    expect(resolvedErrors).toHaveLength(0);
    expect(result.errors).toBeDefined();
    expect(result.errors?.[0].message).toContain('This error is not handled by the plugin');
  });

  it('calls onResolvedError for handled errors but not unhandled ones', async () => {
    const resolvedErrors: Error[] = [];

    class HandledError extends Error {
      constructor(message: string) {
        super(message);
        this.name = 'HandledError';
      }
    }

    class UnhandledError extends Error {
      constructor(message: string) {
        super(message);
        this.name = 'UnhandledError';
      }
    }

    const testBuilder = new SchemaBuilder<{}>({
      plugins: [ErrorPlugin],
      errors: {
        defaultTypes: [],
        onResolvedError: (error) => {
          resolvedErrors.push(error);
        },
      },
    });

    testBuilder.objectType(HandledError, {
      name: 'HandledError',
      fields: (t) => ({
        message: t.exposeString('message'),
      }),
    });

    testBuilder.queryType({
      fields: (t) => ({
        handledField: t.field({
          type: 'String',
          errors: {
            types: [HandledError],
          },
          resolve: () => {
            throw new HandledError('This error is handled');
          },
        }),
        unhandledField: t.field({
          type: 'String',
          errors: {
            types: [HandledError], // Only catching HandledError
          },
          resolve: () => {
            throw new UnhandledError('This error is not handled');
          },
        }),
      }),
    });

    const testSchema = testBuilder.toSchema();

    // Test handled error
    const handledResult = await execute({
      schema: testSchema,
      document: gql`
        query {
          handledField {
            __typename
            ... on HandledError {
              message
            }
          }
        }
      `,
      contextValue: {},
    });

    expect(resolvedErrors).toHaveLength(1);
    expect(resolvedErrors[0]).toBeInstanceOf(HandledError);
    expect(resolvedErrors[0].message).toBe('This error is handled');
    expect(handledResult.errors).toBeUndefined();

    const unhandledResult = await execute({
      schema: testSchema,
      document: gql`
        query {
          unhandledField {
            __typename
          }
        }
      `,
      contextValue: {},
    });

    expect(resolvedErrors).toHaveLength(1);
    expect(unhandledResult.errors).toBeDefined();
    expect(unhandledResult.errors?.[0].message).toContain('This error is not handled');
  });
});
