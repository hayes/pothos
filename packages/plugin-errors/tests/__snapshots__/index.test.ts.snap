// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`errors plugin > generates expected schema 1`] = `
"type BaseError implements Error {
  message: String
}

type CreateUserSuccess {
  id: String
  name: String
}

type DirectResult {
  id: ID
}

interface Error {
  message: String
}

type Extended2Error implements Error {
  code: Int
  message: String
}

type ExtendedError implements Error {
  code: Int
  message: String
}

type InputValidationError implements Error {
  issues: [InputValidationIssue!]
  message: String
}

type InputValidationIssue {
  message: String
  path: [String!]
}

type LengthError implements Error {
  message: String
  minLength: Int
}

type Mutation {
  testErrorUnion(action: String!, shouldFail: String): MutationTestErrorUnionResult
  testErrorUnionCustomResolveType(type: String!): MutationTestErrorUnionCustomResolveTypeResult
  testErrorUnionList(includeError: Boolean!): [MutationTestErrorUnionListItemResult!]
  testErrorUnionListAsyncIterable(includeError: Boolean!, throwError: Boolean!): [MutationTestErrorUnionListAsyncIterableItemResult!]
  testErrorUnionListCustomResolveType(includeError: Boolean!): [MutationTestErrorUnionListCustomResolveTypeItemResult!]
  testErrorUnionListIterable(includeError: Boolean!): [MutationTestErrorUnionListIterableItemResult!]
  testErrorUnionListThrow(shouldThrow: Boolean!): [MutationTestErrorUnionListThrowItemResult!]
  testErrorUnionListWithErrors(includeItemError: Boolean!, throwNotFound: Boolean): MutationTestErrorUnionListWithErrorsResult
  testErrorUnionNullable(returnNull: Boolean!): MutationTestErrorUnionNullableResult
  testErrorUnionThrow(shouldFail: String!): MutationTestErrorUnionThrowResult
  testErrorUnionThrowNonError: MutationTestErrorUnionThrowNonErrorResult
  testErrorUnionWithDefaultTypes(shouldThrowDefault: Boolean!): MutationTestErrorUnionWithDefaultTypesResult
}

union MutationTestErrorUnionCustomResolveTypeResult = BaseError | CreateUserSuccess | NotFoundError | ValidationError

union MutationTestErrorUnionListAsyncIterableItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListCustomResolveTypeItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListIterableItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListThrowItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListWithErrorsItemResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionListWithErrorsResult = BaseError | MutationTestErrorUnionListWithErrorsSuccess | NotFoundError

type MutationTestErrorUnionListWithErrorsSuccess {
  data: [MutationTestErrorUnionListWithErrorsItemResult!]!
}

union MutationTestErrorUnionNullableResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionResult = BaseError | CreateUserSuccess | NotFoundError | UpdateUserSuccess | ValidationError

union MutationTestErrorUnionThrowNonErrorResult = BaseError | CreateUserSuccess | ValidationError

union MutationTestErrorUnionThrowResult = BaseError | CreateUserSuccess | NotFoundError | ValidationError

union MutationTestErrorUnionWithDefaultTypesResult = BaseError | CreateUserSuccess

type NotFoundError implements Error {
  message: String
}

type Query {
  asyncItemErrors(error: Boolean): QueryAsyncItemErrorsResult
  directResult(shouldThrow: Boolean): QueryDirectResultResult
  errorReturned(shouldFail: Boolean): QueryErrorReturnedResult
  extendedError(throw: String): QueryExtendedErrorResult
  extendedErrorList(throw: String): QueryExtendedErrorListResult
  fieldWithValidation(string: String): QueryFieldWithValidationResult
  hello(name: String!): QueryHelloResult
  helloWithMinLength(name: String!): QueryHelloWithMinLengthResult
  itemErrors: [QueryItemErrorsItemResult]!
  itemErrorsDirectResult: [QueryItemErrorsDirectResultItemResult]!
  itemErrorsWithFieldErrors(shouldThrow: Boolean): QueryItemErrorsWithFieldErrorsResult!
  simpleError(throw: Boolean): QuerySimpleErrorResult
  validation2(stringList: [String!]): QueryValidation2Result
}

union QueryAsyncItemErrorsItemResult = BaseError | QueryAsyncItemErrorsItemSuccess

type QueryAsyncItemErrorsItemSuccess {
  data: DirectResult!
}

union QueryAsyncItemErrorsResult = BaseError | QueryAsyncItemErrorsSuccess

type QueryAsyncItemErrorsSuccess {
  data: [QueryAsyncItemErrorsItemResult!]!
}

union QueryDirectResultResult = BaseError | DirectResult

union QueryErrorReturnedResult = BaseError | QueryErrorReturnedSuccess | ValidationError

type QueryErrorReturnedSuccess {
  data: String!
}

union QueryExtendedErrorListResult = BaseError | Extended2Error | ExtendedError | QueryExtendedErrorListSuccess

type QueryExtendedErrorListSuccess {
  data: [String]!
}

union QueryExtendedErrorResult = BaseError | Extended2Error | ExtendedError | QueryExtendedErrorSuccess

type QueryExtendedErrorSuccess {
  data: String!
}

union QueryFieldWithValidationResult = BaseError | InputValidationError | QueryFieldWithValidationSuccess

type QueryFieldWithValidationSuccess {
  result: Boolean!
}

union QueryHelloResult = BaseError | QueryHelloSuccess

type QueryHelloSuccess {
  data: String!
}

union QueryHelloWithMinLengthResult = BaseError | LengthError | QueryHelloWithMinLengthSuccess

type QueryHelloWithMinLengthSuccess {
  data: String!
}

union QueryItemErrorsDirectResultItemResult = BaseError | DirectResult

union QueryItemErrorsItemResult = BaseError | QueryItemErrorsItemSuccess

type QueryItemErrorsItemSuccess {
  data: DirectResult!
}

union QueryItemErrorsWithFieldErrorsItemResult = BaseError | QueryItemErrorsWithFieldErrorsItemSuccess

type QueryItemErrorsWithFieldErrorsItemSuccess {
  data: DirectResult!
}

union QueryItemErrorsWithFieldErrorsResult = BaseError | QueryItemErrorsWithFieldErrorsSuccess

type QueryItemErrorsWithFieldErrorsSuccess {
  data: [QueryItemErrorsWithFieldErrorsItemResult]!
}

union QuerySimpleErrorResult = BaseError | QuerySimpleErrorSuccess

type QuerySimpleErrorSuccess {
  data: String!
}

union QueryValidation2Result = BaseError | InputValidationError | QueryValidation2Success

type QueryValidation2Success {
  result: Boolean!
}

type Subscription {
  test(errorInIterable: Boolean, errorOnResolve: Boolean, errorOnSubscribe: Boolean, returnNull: Boolean): SubscriptionTestResult
}

union SubscriptionTestResult = BaseError | SubscriptionTestSuccess

type SubscriptionTestSuccess {
  data: Int!
}

type UpdateUserSuccess {
  id: String
  updatedFields: [String!]
}

type ValidationError implements Error {
  field: String
  message: String
}"
`;

exports[`errors plugin > supports generating custom names 1`] = `
"type AsyncItemErrors_CUSTOM_ITEM_RESULT_NAME {
  data: DirectResult!
}

union AsyncItemErrors_CUSTOM_ITEM_UNION_NAME = AsyncItemErrors_CUSTOM_ITEM_RESULT_NAME | BaseError

type AsyncItemErrors_CUSTOM_RESULT_NAME {
  data: [AsyncItemErrors_CUSTOM_ITEM_UNION_NAME!]!
}

union AsyncItemErrors_CUSTOM_UNION_NAME = AsyncItemErrors_CUSTOM_RESULT_NAME | BaseError

type BaseError implements Error {
  message: String
}

type CreateUserSuccess {
  id: String
  name: String
}

type DirectResult {
  id: ID
}

union DirectResult_CUSTOM_UNION_NAME = BaseError | DirectResult

interface Error {
  message: String
}

type ErrorReturned_CUSTOM_RESULT_NAME {
  data: String!
}

union ErrorReturned_CUSTOM_UNION_NAME = BaseError | ErrorReturned_CUSTOM_RESULT_NAME | ValidationError

type Extended2Error implements Error {
  code: Int
  message: String
}

type ExtendedError implements Error {
  code: Int
  message: String
}

type ExtendedErrorList_CUSTOM_RESULT_NAME {
  data: [String]!
}

union ExtendedErrorList_CUSTOM_UNION_NAME = BaseError | Extended2Error | ExtendedError | ExtendedErrorList_CUSTOM_RESULT_NAME

type ExtendedError_CUSTOM_RESULT_NAME {
  data: String!
}

union ExtendedError_CUSTOM_UNION_NAME = BaseError | Extended2Error | ExtendedError | ExtendedError_CUSTOM_RESULT_NAME

type FieldWithValidation_CUSTOM_RESULT_NAME {
  result: Boolean!
}

union FieldWithValidation_CUSTOM_UNION_NAME = BaseError | FieldWithValidation_CUSTOM_RESULT_NAME | InputValidationError

type HelloWithMinLength_CUSTOM_RESULT_NAME {
  data: String!
}

union HelloWithMinLength_CUSTOM_UNION_NAME = BaseError | HelloWithMinLength_CUSTOM_RESULT_NAME | LengthError

type Hello_CUSTOM_RESULT_NAME {
  data: String!
}

union Hello_CUSTOM_UNION_NAME = BaseError | Hello_CUSTOM_RESULT_NAME

type InputValidationError implements Error {
  issues: [InputValidationIssue!]
  message: String
}

type InputValidationIssue {
  message: String
  path: [String!]
}

union ItemErrorsDirectResult_CUSTOM_ITEM_UNION_NAME = BaseError | DirectResult

type ItemErrorsWithFieldErrors_CUSTOM_ITEM_RESULT_NAME {
  data: DirectResult!
}

union ItemErrorsWithFieldErrors_CUSTOM_ITEM_UNION_NAME = BaseError | ItemErrorsWithFieldErrors_CUSTOM_ITEM_RESULT_NAME

type ItemErrorsWithFieldErrors_CUSTOM_RESULT_NAME {
  data: [ItemErrorsWithFieldErrors_CUSTOM_ITEM_UNION_NAME]!
}

union ItemErrorsWithFieldErrors_CUSTOM_UNION_NAME = BaseError | ItemErrorsWithFieldErrors_CUSTOM_RESULT_NAME

type ItemErrors_CUSTOM_ITEM_RESULT_NAME {
  data: DirectResult!
}

union ItemErrors_CUSTOM_ITEM_UNION_NAME = BaseError | ItemErrors_CUSTOM_ITEM_RESULT_NAME

type LengthError implements Error {
  message: String
  minLength: Int
}

type Mutation {
  testErrorUnion(action: String!, shouldFail: String): TestErrorUnion_CUSTOM_UNION_NAME
  testErrorUnionCustomResolveType(type: String!): TestErrorUnionCustomResolveType_CUSTOM_UNION_NAME
  testErrorUnionList(includeError: Boolean!): [TestErrorUnionList_CUSTOM_ITEM_UNION_NAME!]
  testErrorUnionListAsyncIterable(includeError: Boolean!, throwError: Boolean!): [TestErrorUnionListAsyncIterable_CUSTOM_ITEM_UNION_NAME!]
  testErrorUnionListCustomResolveType(includeError: Boolean!): [TestErrorUnionListCustomResolveType_CUSTOM_ITEM_UNION_NAME!]
  testErrorUnionListIterable(includeError: Boolean!): [TestErrorUnionListIterable_CUSTOM_ITEM_UNION_NAME!]
  testErrorUnionListThrow(shouldThrow: Boolean!): [TestErrorUnionListThrow_CUSTOM_ITEM_UNION_NAME!]
  testErrorUnionListWithErrors(includeItemError: Boolean!, throwNotFound: Boolean): TestErrorUnionListWithErrors_CUSTOM_UNION_NAME
  testErrorUnionNullable(returnNull: Boolean!): TestErrorUnionNullable_CUSTOM_UNION_NAME
  testErrorUnionThrow(shouldFail: String!): TestErrorUnionThrow_CUSTOM_UNION_NAME
  testErrorUnionThrowNonError: TestErrorUnionThrowNonError_CUSTOM_UNION_NAME
  testErrorUnionWithDefaultTypes(shouldThrowDefault: Boolean!): TestErrorUnionWithDefaultTypes_CUSTOM_UNION_NAME
}

type NotFoundError implements Error {
  message: String
}

type Query {
  asyncItemErrors(error: Boolean): AsyncItemErrors_CUSTOM_UNION_NAME
  directResult(shouldThrow: Boolean): DirectResult_CUSTOM_UNION_NAME
  errorReturned(shouldFail: Boolean): ErrorReturned_CUSTOM_UNION_NAME
  extendedError(throw: String): ExtendedError_CUSTOM_UNION_NAME
  extendedErrorList(throw: String): ExtendedErrorList_CUSTOM_UNION_NAME
  fieldWithValidation(string: String): FieldWithValidation_CUSTOM_UNION_NAME
  hello(name: String!): Hello_CUSTOM_UNION_NAME
  helloWithMinLength(name: String!): HelloWithMinLength_CUSTOM_UNION_NAME
  itemErrors: [ItemErrors_CUSTOM_ITEM_UNION_NAME]!
  itemErrorsDirectResult: [ItemErrorsDirectResult_CUSTOM_ITEM_UNION_NAME]!
  itemErrorsWithFieldErrors(shouldThrow: Boolean): ItemErrorsWithFieldErrors_CUSTOM_UNION_NAME!
  simpleError(throw: Boolean): SimpleError_CUSTOM_UNION_NAME
  validation2(stringList: [String!]): Validation2_CUSTOM_UNION_NAME
}

type SimpleError_CUSTOM_RESULT_NAME {
  data: String!
}

union SimpleError_CUSTOM_UNION_NAME = BaseError | SimpleError_CUSTOM_RESULT_NAME

type Subscription {
  test(errorInIterable: Boolean, errorOnResolve: Boolean, errorOnSubscribe: Boolean, returnNull: Boolean): Test_CUSTOM_UNION_NAME
}

union TestErrorUnionCustomResolveType_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess | NotFoundError | ValidationError

union TestErrorUnionListAsyncIterable_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionListCustomResolveType_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionListIterable_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionListThrow_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionListWithErrors_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

type TestErrorUnionListWithErrors_CUSTOM_RESULT_NAME {
  data: [TestErrorUnionListWithErrors_CUSTOM_ITEM_UNION_NAME!]!
}

union TestErrorUnionListWithErrors_CUSTOM_UNION_NAME = BaseError | NotFoundError | TestErrorUnionListWithErrors_CUSTOM_RESULT_NAME

union TestErrorUnionList_CUSTOM_ITEM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionNullable_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionThrowNonError_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess | ValidationError

union TestErrorUnionThrow_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess | NotFoundError | ValidationError

union TestErrorUnionWithDefaultTypes_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess

union TestErrorUnion_CUSTOM_UNION_NAME = BaseError | CreateUserSuccess | NotFoundError | UpdateUserSuccess | ValidationError

type Test_CUSTOM_RESULT_NAME {
  data: Int!
}

union Test_CUSTOM_UNION_NAME = BaseError | Test_CUSTOM_RESULT_NAME

type UpdateUserSuccess {
  id: String
  updatedFields: [String!]
}

type Validation2_CUSTOM_RESULT_NAME {
  result: Boolean!
}

union Validation2_CUSTOM_UNION_NAME = BaseError | InputValidationError | Validation2_CUSTOM_RESULT_NAME

type ValidationError implements Error {
  field: String
  message: String
}"
`;
